### PR TITLE
type promote clamp

### LINKED ISSUE #76630: `torch.clamp` type promotes strangely
### 🐛 Describe the bug

`torch.clamp` broadcast strangely that its return tensor has the same dtype as `input` argument. However, `torch.max` and `torch.min` will broadcast the return tensor to the most precise one. For example,

```python
import torch
input = torch.rand([5], dtype=torch.float32)
other = torch.rand([5, 5], dtype=torch.float64)

r1 = torch.maximum(input, other)
r2 = torch.clamp(input, min=other)
print(r1.dtype, r2.dtype)
# torch.float64 torch.float32
```

In the documentation of `torch.clamp`, `y = min(max(x, min_value), max_value)`, so I think it should the same broadcasting behavior as `torch.max` and `torch.min`

### Versions

pytorch: 1.10.0

cc @nairbv @mruberry @ezyang @ngimel

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #76630
When clamp(Tensor, Tensor) is structured, big parts of this PR won't be needed, but for now let's fix type promotion to make behavior more regular. 
Some special-casing is done in hardtanh to preserve legacy behavior of not promoting to boundaries. 
Also, relu is explicitly disabled on bool inputs to avoid accidental promotion to integer type. 
BC-breaking note:
Previously, `min` and `max` arguments in clamp did not participate in type promotion, which made it inconsistent with `minimum` and `maximum` operations. This PR (and PRs porting clamp_min/clamp_max to structured) make `min` and `max` arguments participate in type promotion.

### CODE REVIEW

File: aten/src/ATen/native/TypeProperties.h [hunk 1]
```diff
@@ -12,6 +12,7 @@ struct ResultTypeState {
 };
 
 TORCH_API ResultTypeState update_result_type_state(const Tensor& tensor, const ResultTypeState& in_state);
+TORCH_API ResultTypeState update_result_type_state(const Scalar& scalar, const ResultTypeState& in_state);
```
Comment (mruberry) [inline aten/src/ATen/native/TypeProperties.h:15]: Nice addition

**Review by mruberry — APPROVED** [review]
Cool! Stamped!

**Review by albanD — COMMENTED** [review]
Should this be marked as BC-breaking?

**Comment by ngimel** [comment]
@alband good point

**Pull request closed without merging**